server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java

```diff
@@ -125,13 +126,24 @@ protected void asyncShardOperation(GetRequest request, ShardId shardId, ActionLi
         IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
         IndexShard indexShard = indexService.getShard(shardId.id());
         if (indexShard.routingEntry().isPromotableToPrimary() == false) {
+            // TODO: remove assertion (ES-9563)
             assert indexShard.indexSettings().isFastRefresh() == false
-                : "a search shard should not receive a TransportGetAction for an index with fast refresh";
+                || clusterService.state().getMinTransportVersion().onOrAfter(TransportVersions.FAST_REFRESH_RCO)
+                : "a search shard in a cluster with min transport version "
+                    + clusterService.state().getMinTransportVersion()
+                    + " (before FAST_REFRESH_RCO) should not receive a TransportGetAction for an index with fast refresh";
             handleGetOnUnpromotableShard(request, indexShard, listener);
             return;
         }
-        assert DiscoveryNode.isStateless(clusterService.getSettings()) == false || indexShard.indexSettings().isFastRefresh()
-            : "in Stateless a promotable to primary shard can receive a TransportGetAction only if an index has the fast refresh setting";
+        // TODO: adapt assertion to assert only that it is not stateless (ES-9563)
+        assert DiscoveryNode.isStateless(clusterService.getSettings()) == false
+            || (indexShard.indexSettings().isFastRefresh()
+                && clusterService.state().getMinTransportVersion().before(TransportVersions.FAST_REFRESH_RCO))
```

Comment:
Similarly, here, a GET message for a fast refresh shard could be sent from a node where the min-transport-version had not advanced but then received on this search node after it advances, making this assertion unsafe.